File: MobileTrailer/ServersViewController.m

@interface ServersViewController ()
{
	NSManagedObjectID *selectedServerId;
	NSMutableArray *allServers;
	NSDateFormatter *resetFormatter;
}
@end

@implementation ServersViewController

- (IBAction)done:(UIBarButtonItem *)sender
{
	if(app.preferencesDirty) [app startRefresh];
	[self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.clearsSelectionOnViewWillAppear = YES;
	resetFormatter = [[NSDateFormatter alloc] init];
	resetFormatter.doesRelativeDateFormatting = YES;
	resetFormatter.timeStyle = NSDateFormatterShortStyle;
	resetFormatter.dateStyle = NSDateFormatterShortStyle;

	[[NSNotificationCenter defaultCenter] addObserver:self.tableView
											 selector:@selector(reloadData)
												 name:REFRESH_ENDED_NOTIFICATION
											   object:nil];
}

- (void)dealloc
{
	[[NSNotificationCenter defaultCenter] removeObserver:self.tableView];
}

- (void)viewWillAppear:(BOOL)animated
{
	[super viewWillAppear:animated];
	allServers = [[ApiServer allApiServersInMoc:app.dataManager.managedObjectContext] mutableCopy];
	[self.tableView reloadData];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView { return 1; }

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
	return allServers.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ServerCell" forIndexPath:indexPath];
	ApiServer *a = allServers[indexPath.row];
	if(a.authToken.length==0)
	{
		cell.textLabel.textColor = [UIColor redColor];
		cell.textLabel.text = [a.label stringByAppendingString:@" (needs token!)"];
	}
	else if(!a.lastSyncSucceeded.boolValue)
	{
		cell.textLabel.textColor = [UIColor redColor];
		cell.textLabel.text = [a.label stringByAppendingString:@" (last sync failed)"];
	}
	else
	{
		cell.textLabel.textColor = [UIColor darkTextColor];
		cell.textLabel.text = a.label;
	}
	if(a.requestsLimit.doubleValue==0.0)
	{
		cell.detailTextLabel.text = nil;
	}
	else
	{
		double total = a.requestsLimit.doubleValue;
		double used = total-a.requestsRemaining.doubleValue;
		if(a.resetDate)
		{
			cell.detailTextLabel.text = [NSString stringWithFormat:@"%.01f%% API used (%.0f / %.0f requests)\nNext reset: %@",
										 100*used/total, used, total, [resetFormatter stringFromDate:a.resetDate]];
		}
		else
		{
			cell.detailTextLabel.text = [NSString stringWithFormat:@"%.01f%% API used (%.0f / %.0f requests)",
										 100*used/total, used, total];
		}
	}
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath { return YES; }

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete)
	{
		ApiServer *a = allServers[indexPath.row];
		[allServers removeObjectAtIndex:indexPath.row];
		[app.dataManager.managedObjectContext deleteObject:a];
		[app.dataManager saveDB];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	ApiServer *a = allServers[indexPath.row];
	selectedServerId = a.objectID;
	[self performSegueWithIdentifier:@"editServer" sender:self];
}

- (IBAction)newServer:(UIBarButtonItem *)sender
{
	[self performSegueWithIdentifier:@"editServer" sender:self];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
	ServerDetailViewController *sd = [segue destinationViewController];
	sd.serverId = selectedServerId;
	selectedServerId = nil;
}

@end

File: PocketTrailer/ServersViewController.swift
import CoreData
import UIKit

final class ServersViewController: UITableViewController {
    private var selectedServerId: NSManagedObjectID?
    private var allServers = [ApiServer]()

    @IBOutlet var apiSwitch: UIBarButtonItem!

    @IBAction private func doneSelected() {
        dismiss(animated: true)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        clearsSelectionOnViewWillAppear = true
        NotificationCenter.default.addObserver(tableView!, selector: #selector(UITableView.reloadData), name: .RefreshEnded, object: nil)
        updateApiLabel()
        navigationItem.largeTitleDisplayMode = .automatic
    }

    private func updateApiLabel() {
        apiSwitch.title = Settings.useV4API ? "Using v4 API" : "Using v3 API"
    }

    deinit {
        if let tableView {
            NotificationCenter.default.removeObserver(tableView)
        }
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        allServers = ApiServer.allApiServers(in: DataManager.main)
        tableView.reloadData()
    }

    override func tableView(_: UITableView, numberOfRowsInSection _: Int) -> Int {
        allServers.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "ServerCell", for: indexPath)
        if let T = cell.textLabel, let D = cell.detailTextLabel {
            let a = allServers[indexPath.row]
            if a.authToken.isEmpty {
                T.textColor = .appRed
                T.text = "\(a.label.orEmpty) (needs token!)"
            } else if !a.lastSyncSucceeded {
                T.textColor = .appRed
                T.text = "\(a.label.orEmpty) (last sync failed)"
            } else {
                T.textColor = UIColor.label
                T.text = a.label
            }
            if a.requestsLimit == 0 {
                D.text = nil
            } else {
                let total = Double(a.requestsLimit)
                let used = Double(total - Double(a.requestsRemaining))
                if a.resetDate != nil {
                    D.text = String(format: "%.01f%% API used (%.0f / %.0f requests)\nNext reset: %@", 100 * used / total, used, total, shortDateFormatter.string(from: a.resetDate!))
                } else {
                    D.text = String(format: "%.01f%% API used (%.0f / %.0f requests)", 100 * used / total, used, total)
                }
            }
        }
        return cell
    }

    override func tableView(_: UITableView, canEditRowAt _: IndexPath) -> Bool {
        true
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == UITableViewCell.EditingStyle.delete {
            let a = allServers[indexPath.row]
            allServers.remove(at: indexPath.row)
            DataManager.main.delete(a)
            Task {
                await DataManager.saveDB()
                tableView.deleteRows(at: [indexPath], with: .fade)
            }
        }
    }

    override func tableView(_: UITableView, didSelectRowAt indexPath: IndexPath) {
        let a = allServers[indexPath.row]
        selectedServerId = a.objectID
        performSegue(withIdentifier: "editServer", sender: self)
    }

    @IBAction private func newServer() {
        performSegue(withIdentifier: "editServer", sender: self)
    }

    override func prepare(for segue: UIStoryboardSegue, sender _: Any?) {
        if let sd = segue.destination as? ServerDetailViewController {
            sd.serverLocalId = selectedServerId
            selectedServerId = nil
        }
    }

    @IBAction private func apiToggleSelected(_ sender: UIBarButtonItem) {
        let a = UIAlertController(title: "API Options", message: Settings.useV4APIHelp, preferredStyle: .actionSheet)
        a.addAction(UIAlertAction(title: "Use legacy v3 API", style: .default) { [weak self] _ in
            Settings.useV4API = false
            self?.apiChanged()
        })
        a.addAction(UIAlertAction(title: "Use v4 API", style: .default) { [weak self] _ in
            if let error = API.canUseV4API(for: DataManager.main) {
                showMessage(Settings.v4title, error)
            } else {
                Settings.useV4API = true
                self?.apiChanged()
            }
        })
        a.addAction(UIAlertAction(title: "v4 API Options…", style: .default) { [weak self] _ in
            if let error = API.canUseV4API(for: DataManager.main) {
                showMessage(Settings.v4title, error)
            } else {
                self?.performSegue(withIdentifier: "apiOptions", sender: nil)
            }
        })
        a.addAction(UIAlertAction(title: "Log Monitor…", style: .default) { [weak self] _ in
            self?.performSegue(withIdentifier: "liveLog", sender: nil)
        })
        a.addAction(UIAlertAction(title: "Hidden Items…", style: .default) { [weak self] _ in
            self?.performSegue(withIdentifier: "hiddenItems", sender: nil)
        })
        a.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        present(a, animated: true)
        a.popoverPresentationController?.barButtonItem = sender
    }

    private func apiChanged() {
        updateApiLabel()
        for apiServer in ApiServer.allApiServers(in: DataManager.main) {
            apiServer.deleteEverything()
            apiServer.resetSyncState()
        }
        Task {
            await DataManager.saveDB()
        }
    }

    @IBAction private func resyncEverythingSelected(_ sender: UIBarButtonItem) {
        let a = UIAlertController(title: sender.title, message: Settings.reloadAllDataHelp, preferredStyle: .actionSheet)
        a.addAction(UIAlertAction(title: sender.title, style: .destructive) { [weak self] _ in
            self?.performFullReload()
        })
        a.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        present(a, animated: true)
        a.popoverPresentationController?.barButtonItem = sender
    }

    private func performFullReload() {
        for a in ApiServer.allApiServers(in: DataManager.main) {
            a.deleteEverything()
            a.resetSyncState()
        }
        Task {
            await DataManager.saveDB()
            await DataManager.postProcessAllItems(in: DataManager.main, settings: Settings.cache)
            _ = await app.startRefresh()
        }
    }
}
